code_generation/configuration.py

```diff
@@ -0,0 +1,526 @@
+
+log = logging.getLogger(__name__)
+# type aliases
+ScopeSet = Set[str]
+ScopeList = Union[str, List[str]]
+EraList = Union[str, List[str]]
+ChannelList = Union[str, List[str]]
+ShiftList = Union[str, List[str]]
+SamplesList = Union[str, List[str]]
+TConfiguration = Dict[
+    str,
+    Union[
+        List[Union[str, int, float, bool, Dict[Any, Any]]],
+        str,
+        int,
+        float,
+        bool,
+        EraModifier,
+        SampleModifier,
+        Dict[Any, Any],
+    ],
+]
+
+
+class Configuration(object):
+    """
+    Configuration class for for the CROWN configuration. This class
+    holds all parts of the configuration, from the sample, era, channel,
+    and systematics, to the output. All modifications to the configuration should be done through this class.
+    """
+
+    def __init__(
+        self,
+        era: str,
+        sample: str,
+        channels: ChannelList,
+        shifts: ShiftList,
+        available_sample_types: SamplesList,
+        available_eras: EraList,
+        available_channels: ChannelList,
+    ):
+        """
+
+        Initiation of a new configuration.
+
+        Args:
+            era: The era of the sample.
+            sample: The sample type of the sample.
+            channels: The channels to be used in the configuration. Each channel is considered a scope in CROWN.
+            By default, the global scope will always be added and run first, all other scopes will be run in
+            parallel as children of the global scope.
+            shifts: The systematics to be used in the configuration.
+            available_sample_types: The available sample types.
+            available_eras: The available eras.
+            available_channels: The available channels.
+
+        """
+        self.era = era
+        self.sample = sample
+        self.channels = set(channels)
+        self.selected_shifts = set(shifts)
+        self.available_sample_types = set(available_sample_types)
+        self.available_eras = set(available_eras)
+        self.available_channels = set(available_channels)
+        self.available_outputs: QuantitiesStore = {}
+        self.global_scope = "global"
+
+        self.producers: TProducerStore = {}
+        self.scopes: List[str] = []
+        self.outputs: QuantitiesStore = {}
+        self.shifts: Dict[str, Dict[str, TConfiguration]] = {}
+        self.rules: Set[ProducerRule] = set()
+        self.config_parameters: Dict[str, TConfiguration] = {}
+
+        self.setup_defaults()
+
+    def validate_channels(self) -> None:
+        """
+        Function to validate the selected channels. If the channel is not available, an error is raised.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        missing_channels = self.channels - self.available_channels
+        if len(missing_channels) > 0:
+            raise ChannelConfigurationError(self.channels, missing_channels)
```

Comment:
Shouldn't this be 
`raise ChannelConfigurationError(missing_channels, self.available_channels)`
based on the definition [here](https://github.com/KIT-CMS/CROWN/blob/eff8756bcf070769078e2c7ed6d840eaac37f1e1/code_generation/exceptions.py#L34) ?

Reply:
true !